internal/graph/check.go

```diff
@@ -141,313 +139,226 @@ type CheckHandlerFunc func(ctx context.Context) (*ResolveCheckResponse, error)
-	limiter := make(chan struct{}, concurrencyLimit)
+// runHandler safely executes a CheckHandlerFunc, recovers from any panics,
+// and returns the result as a checkOutcome.
+func runHandler(ctx context.Context, handler CheckHandlerFunc) checkOutcome {
+	var res *ResolveCheckResponse
+	var err error
 
-	var wg conc.WaitGroup
-
-	checker := func(fn CheckHandlerFunc) {
-		defer func() {
-			<-limiter
-		}()
-
-		resolved := make(chan checkOutcome, 1)
-
-		if ctx.Err() != nil {
-			resultChan <- checkOutcome{nil, ctx.Err()}
-			return
-		}
-
-		wg.Go(func() {
-			defer close(resolved)
-
-			resp, err := fn(ctx)
-			resolved <- checkOutcome{resp, err}
-		})
-
-		select {
-		case <-ctx.Done():
-			return
-		case res := <-resolved:
-			resultChan <- res
-		}
-	}
-
-	wg.Go(func() {
-	outer:
-		for _, handler := range handlers {
-			fn := handler // capture loop var
-
-			select {
-			case limiter <- struct{}{}:
-				wg.Go(func() {
-					checker(fn)
-				})
-			case <-ctx.Done():
-				break outer
-			}
-		}
+	recoveredErr := panics.Try(func() {
+		res, err = handler(ctx)
 	})
-
-	return func() error {
-		recoveredError := wg.WaitAndRecover()
-		close(limiter)
-
-		if recoveredError != nil {
-			return fmt.Errorf("%w: %s", ErrPanic, recoveredError.AsError())
-		}
-
-		return nil
+	if recoveredErr != nil {
+		err = fmt.Errorf("%w: %s", ErrPanic, recoveredErr.AsError())
 	}
+
+	return checkOutcome{res, err}
 }
 
 // union implements a CheckFuncReducer that requires any of the provided CheckHandlerFunc to resolve
 // to an allowed outcome. The first allowed outcome causes premature termination of the reducer.
 func union(ctx context.Context, concurrencyLimit int, handlers ...CheckHandlerFunc) (resp *ResolveCheckResponse, err error) {
-	ctx, cancel := context.WithCancel(ctx)
-	resultChan := make(chan checkOutcome, len(handlers))
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	drain := resolver(ctx, concurrencyLimit, resultChan, handlers...)
+	pool := concurrency.NewPool(cancellableCtx, concurrencyLimit)
+	out := make(chan checkOutcome, len(handlers))
 
-	defer func() {
-		cancel()
-		drainErr := drain()
-		if drainErr != nil {
-			err = drainErr
-			resp = nil
-		}
-		close(resultChan)
+	for _, handler := range handlers {
+		h := handler
+		pool.Go(func(ctx context.Context) error {
+			concurrency.TrySendThroughChannel(cancellableCtx, runHandler(ctx, h), out)
+			return nil
+		})
+	}
+
+	go func() {
+		_ = pool.Wait()
+		close(out)
 	}()
 
-	var elErr error
-	var cycleDetected bool
+	var finalErr error
+	finalResult := &ResolveCheckResponse{Allowed: false}
+
 	for i := 0; i < len(handlers); i++ {
 		select {
-		case result := <-resultChan:
-			if result.err != nil {
-				elErr = result.err
-				continue
+		case <-ctx.Done():
+			return nil, ctx.Err()
+
+		case outcome, ok := <-out:
+			if !ok {
+				break
 			}
 
-			if result.resp.GetCycleDetected() {
-				cycleDetected = true
+			if outcome.err != nil {
```

Comment:
This is slightly different than the intersection case.  In the intersection case, we store the first error but here we store the last error.  Is this intended?